Image signature verification workflow test

mfojtik opened the pull request:
@miminar @soltysh as promised

This will basically:

* Take the `origin:latest` image (which should be the local image we built using `make release`
* Add `skopeo` into it and generate dummy GPG keys and commit it as "signer" image
* Use that image for a "sign&push" job (I know Maciej, I should use Job this ;-)
* Once the `signed:latest` is present, check the signature using describe
* Use the "signer" image once again to run `oc adm verify-image-signature` and verify the identity of the image
* Check the image is now "verified"

I'm pretty hesitating to add this into "comformance" as the test takes ~1minute to run, thoughts?

mfojtik commented:
[test][testextended][extended:core(image signature workflow)]

miminar commented on test/extended/registry/signature.go:
```diff
@@ -0,0 +1,116 @@
+ENV GNUPGHOME="/var/lib/origin/gnupg"
+RUN yum install -y skopeo && yum clean all
+RUN chmod 0777 -R /var/lib/origin && mkdir -p $GNUPGHOME && chmod 0600 $GNUPGHOME && \
+rm -f /dev/random; ln -sf /dev/urandom /dev/random && \
```
Ugh?

Maybe worth a comment.

mfojtik replied:
@miminar containers lack entropy ;-)

miminar commented on test/extended/registry/signature.go:
```diff
@@ -0,0 +1,116 @@
+		err = exutil.WaitForAnImageStreamTag(oc, oc.Namespace(), "signed", "latest")
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("checking the signature is present on the image and it state is unverified")
```
`s/on/in/` `s/it/its/`

miminar commented on test/extended/registry/signature.go:
```diff
@@ -0,0 +1,116 @@
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("signing the origin-pod:latest image and pushing it into signed:latest")
+		// TODO: Note that the skopeo command is expecting the KUBERNETES_MASTER and the
```
This should either go without `TODO:` or state what needs to be done.

miminar commented on test/extended/registry/signature.go:
```diff
@@ -0,0 +1,116 @@
+		// image-signer and image-auditor roles.
+		signAndPushCmd := fmt.Sprintf(`export KUBERNETES_MASTER=https://$KUBERNETES_SERVICE_HOST:$KUBERNETES_SERVICE_PORT;`+
+			`oc login --token=%s $KUBERNETES_MASTER --certificate-authority=/run/secrets/kubernetes.io/serviceaccount/ca.crt && `+
+			`skopeo --debug --tls-verify=false copy --sign-by joe@foo.bar --dest-creds %s:%s --dest-tls-verify=false docker://docker.io/openshift/origin-pod:latest atomic:%s/%s/signed:latest`,
```
Why is this signing `docker.io/openshift/origin-pod:latest`? Can we use some local image? Is this pulling the image or just a manifest?

mfojtik replied:
@miminar this pod does not have access to local docker (!) it uses skopeo transport. i don't want it to have privileged access.

bparees commented on test/extended/registry/signature.go:
```diff
@@ -0,0 +1,116 @@
+
+	g.It("can push a signed image to openshift registry and verify it", func() {
+		g.By("building an signer image that know how to sign images")
+		_, err := oc.Run("new-build").Args("--docker-image", "openshift/origin:latest", "--to", "signer:latest", "-D", "-").InputString(`FROM openshift/origin:latest
```
you can do it, but not via new-build/new-app.  you'd have to define the build yourself using a dockerimage reference.

soltysh replied:
We should have `oc create buildconfig` :D

soltysh replied:
I've created https://github.com/openshift/origin/issues/14138.

mfojtik replied:
i will create the fixture for now :-)

soltysh commented on test/extended/registry/signature.go:
```diff
@@ -0,0 +1,120 @@
+		// Note that we need to replace the /dev/random with /dev/urandom to get more entropy
+		// into container so we can successfully generate the GPG keypair.
+		g.By("creating dummy GPG key")
+		out, err := pod.Exec("rm -f /dev/random; ln -sf /dev/urandom /dev/random && " +
```
:worried:

mfojtik replied:
@soltysh without this the container will just hang forever as it does not have enough entropy... the same hack is used in the skopeo integration testing...

soltysh replied:
I'm not saying you can't do this, I'm just expressing my level of sadness with this hack, sigh...

soltysh commented on test/extended/util/framework.go:
```diff
@@ -1439,3 +1439,35 @@ func CheckForBuildEvent(client kcoreclient.CoreV1Interface, build *buildapi.Buil
+	waitErr := wait.PollImmediate(1*time.Second, 3*time.Minute, func() (bool, error) {
+		var err error
+		out, err = r.client.Run("exec").Args(r.podName, "--", "/bin/bash", "-c", script).Output()
+		framework.Logf("output:\n%s\n", out)
```
Drop this line, not needed here. If you want to log it, make sure the caller does that for you, since you're returning that output.

mfojtik replied:
sure. (this was for debugging the test only)

miminar commented on test/extended/util/framework.go:
```diff
@@ -1439,3 +1439,34 @@ func CheckForBuildEvent(client kcoreclient.CoreV1Interface, build *buildapi.Buil
+// Exec executes a single command or a script in the running pod. It returns the command
+// output and error if the command finished with non-zero status code or the command took
+// longer then 3 minutes to run.
+func (r *podExecutor) Exec(script string) (string, error) {
```
The godoc could mention that the `script` is a script interpreted by bash.

mfojtik replied:
done

miminar commented on test/extended/registry/signature.go:
```diff
@@ -0,0 +1,120 @@
+package registry
```
IMHO this should belong rather to `test/extended/imageapis`. But I admit it's ambiguous. Tests in the `imageapis` directories work with registry as well. And the same goes for tests in `images` and `builds` directory.

mfojtik replied:
yeah, i don't think either of these packages is a great location for this test, but I would like to have it in registry as the verification fetching manifest from registry and I want to add additional test in (near) future for the signature endpoint in registry (don't want to have the signature test splitted in two packages)

miminar commented on test/extended/registry/signature.go:
```diff
@@ -0,0 +1,120 @@
+	e2e "k8s.io/kubernetes/test/e2e/framework"
+)
+
+var _ = g.Describe("[imageapis] image signature workflow", func() {
```
Since it involves registry, there's no harm adding `[registry]` as well. Sorry I missed that earlier.

mfojtik replied:
sure.

mfojtik commented on test/extended/registry/signature.go:
```diff
@@ -0,0 +1,120 @@
+		// TODO: Fix skopeo to pickup the Kubernetes environment variables (remove the $KUBERNETES_MASTER)
+		g.By("signing the origin-pod:latest image and pushing it into openshift registry")
+		_, err = pod.Exec("KUBERNETES_MASTER=https://$KUBERNETES_SERVICE_HOST:$KUBERNETES_SERVICE_PORT GNUPGHOME=/var/lib/origin/gnupg " +
+			"skopeo --debug --tls-verify=false copy --sign-by joe@foo.bar --dest-creds " + user + ":" + token + " --dest-tls-verify=false docker://docker.io/openshift/origin-pod:latest atomic:" + signedImage)
```
@miminar i have no idea how to plumb the docker socket to skopeo in container... i guess it will make the test pretty complicated...

mfojtik replied:
@miminar also skopeo does not use the docker to transport images, it has custom functions to deal with that.

mfojtik commented:
@miminar success \o/

mfojtik commented:
@miminar @soltysh moved the build into its own fixture (which is btw. cool because we don't have dockerfile test anyway ;-) addressed comments, ptal.

next step will be registry signature endpoint test (part of this test)

mfojtik commented:
[merge]

mfojtik commented:
[test]

mfojtik commented:
flake https://github.com/openshift/origin/issues/14241

[merge]

mfojtik commented:
flake: https://github.com/openshift/origin/issues/14236 (fixed)

[merge]
